File: python/bistring/_infer.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT license.

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional
import unicodedata

from ._alignment import Alignment
from ._bistr import bistr
from ._token import CharacterTokenizer


@dataclass(frozen=True)
class AugmentedChar:
    """
    A single character (grapheme cluster) augmented with extra information.
    """

    top_category: str
    """
    The top-level Unicode category of the char (L, P, Z, etc.).
    """

    category: str
    """
    The specific Unicode category of the char (Lu, Po, Zs, etc.).
    """

    root: str
    """
    The root code point of the grapheme cluster.
    """

    folded: str
    """
    The case-folded form of the char.
    """

    normalized: str
    """
    The Unicode compatibility normalized form of the char.
    """

    original: str
    """
    The original form of the char.
    """

    @classmethod
    def cost_fn(cls, a: Optional[AugmentedChar], b: Optional[AugmentedChar]) -> int:
        """
        The cost function between augmented chars.  Each attribute contributes one "point" towards their distance.
        """

        if a is None or b is None:
            # cost(insert) + cost(delete) (4 + 4) should be more than cost(substitute) (6)
            return 4

        result = 0
        result += int(a.top_category != b.top_category)
        result += int(a.category != b.category)
        result += int(a.root != b.root)
        result += int(a.folded != b.folded)
        result += int(a.normalized != b.normalized)
        result += int(a.original != b.original)
        return result


TOKENIZER = CharacterTokenizer('root')


@dataclass(frozen=True)
class AugmentedString:
    """
    A string augmented with extra information about each character.
    """

    original: str
    """
    The original string.
    """

    chars: List[AugmentedChar]
    """
    The augmented characters of the string.
    """

    alignment: Alignment
    """
    The alignment between the original string and the augmented chars.
    """

    @classmethod
    def augment(cls, original: str) -> AugmentedString:
        normalized = bistr(original).normalize('NFKD')
        folded = bistr(normalized.modified).casefold()
        glyphs = TOKENIZER.tokenize(folded)

        chars = []
        for glyph in glyphs:
            fold_c = glyph.text.modified
            root = fold_c[0]

            norm_slice = folded.alignment.original_slice(glyph.start, glyph.end)
            norm_c = folded.original[norm_slice]

            orig_slice = normalized.alignment.original_slice(norm_slice)
            orig_c = normalized.original[orig_slice]

            cat = unicodedata.category(root)
            top_cat = cat[0]

            chars.append(AugmentedChar(top_cat, cat, root, fold_c, norm_c, orig_c))

        alignment = normalized.alignment
        alignment = alignment.compose(folded.alignment)
        alignment = alignment.compose(glyphs.alignment)
        return cls(original, chars, alignment)


def heuristic_infer(original: str, modified: str) -> bistr:
    """
    Infer the alignment between two strings with a "smart" heuristic.

    We use Unicode normalization and case folding to minimize differences that are due to case, accents, ligatures, etc.
    """

    aug_orig = AugmentedString.augment(original)
    aug_mod = AugmentedString.augment(modified)

    alignment = Alignment.infer(aug_orig.chars, aug_mod.chars, AugmentedChar.cost_fn)
    alignment = aug_orig.alignment.compose(alignment)
    alignment = alignment.compose(aug_mod.alignment.inverse())

    return bistr(original, modified, alignment)

File: js/src/infer.ts
import Alignment, { BiIndex } from "./alignment";
import BiString from "./bistring";

// https://unicode.org/reports/tr44/#GC_Values_Table
const CATEGORIES = [
    "Lu", "Ll", "Lt", "Lm", "Lo",
    "Mn", "Mc", "Me",
    "Nd", "Nl", "No",
    "Pc", "Pd", "Ps", "Pe", "Pi", "Pf", "Po",
    "Sm", "Sc", "Sk", "So",
    "Zs", "Zl", "Zp",
    "Cc", "Cf", "Cs", "Co", "Cn",
];

// TODO: Babel doesn't polyfill this, and the source transformation doesn't catch it
// const CATEGORY_REGEXP = new RegExp(CATEGORIES.map(c => `(\\p{${c}})`).join("|"), "u");
const CATEGORY_REGEXP = /(\p{Lu})|(\p{Ll})|(\p{Lt})|(\p{Lm})|(\p{Lo})|(\p{Mn})|(\p{Mc})|(\p{Me})|(\p{Nd})|(\p{Nl})|(\p{No})|(\p{Pc})|(\p{Pd})|(\p{Ps})|(\p{Pe})|(\p{Pi})|(\p{Pf})|(\p{Po})|(\p{Sm})|(\p{Sc})|(\p{Sk})|(\p{So})|(\p{Zs})|(\p{Zl})|(\p{Zp})|(\p{Cc})|(\p{Cf})|(\p{Cs})|(\p{Co})|(\p{Cn})/u;

function category(cp: string): string {
    const match = cp.match(CATEGORY_REGEXP)!;
    const index = match.indexOf(cp, 1) - 1;
    return CATEGORIES[index];
}

/**
 * A single glyph in a string, augmented with some extra information.
 */
class AugmentedGlyph {
    /** The original form of the glyph. */
    readonly original: string;
    /** The Unicode compatibility normalized form of the glyph. */
    readonly normalized: string;
    /** The uppercase form of the glyph. */
    readonly upper: string;
    /** The root code point of the grapheme cluster. */
    readonly root: string;
    /** The specific Unicode category of the glyph (Lu, Po, Zs, etc.). */
    readonly category: string;
    /** The top-level Unicode category of the glyph (L, P, Z, etc.). */
    readonly topCategory: string;

    constructor(original: string, normalized: string, upper: string) {
        this.original = original;
        this.normalized = normalized;
        this.upper = upper;
        this.root = String.fromCodePoint(upper.codePointAt(0)!);
        this.category = category(this.root);
        this.topCategory = this.category[0];
    }

    static costFn(a?: AugmentedGlyph, b?: AugmentedGlyph) {
        if (!a || !b) {
            // cost(insert) + cost(delete) (4 + 4) should be more than cost(substitute) (6)
            return 4;
        }

        let result = 0;
        result += +(a.original !== b.original);
        result += +(a.normalized !== b.normalized);
        result += +(a.upper !== b.upper);
        result += +(a.root !== b.root);
        result += +(a.category !== b.topCategory);
        result += +(a.topCategory !== b.topCategory);
        return result;
    }
}

/**
 * Not quite as good as UAX #29 grapheme clusters, but we're waiting on Intl.Segmenter.
 */
const GLYPH_REGEXP = /\P{M}\p{M}*|^\p{M}+/gu;

/**
 * A string augmented with some extra information about each glyph.
 */
class AugmentedString {
    /** The original string. */
    readonly original: string;
    /** The augmented glyphs of the string. */
    readonly glyphs: readonly AugmentedGlyph[];
    /** The alignment between the original string and the augmented glyphs. */
    readonly alignment: Alignment;

    constructor(original: string) {
        const normalized = new BiString(original).normalize("NFKD");
        const upper = new BiString(normalized.modified).toUpperCase();

        const glyphs = [];
        const alignment: BiIndex[] = [[0, 0]];
        for (const match of upper.matchAll(GLYPH_REGEXP)) {
            const [o, m] = alignment[alignment.length - 1];

            const upperC = match[0];

            const normBounds = upper.alignment.originalBounds(o, o + upperC.length);
            const normC = upper.original.slice(...normBounds);

            const origBounds = normalized.alignment.originalBounds(normBounds);
            const origC = normalized.original.slice(...origBounds);

            glyphs.push(new AugmentedGlyph(origC, normC, upperC));

            alignment.push([o + normC.length, m + 1]);
        }
        this.original = original;
        this.glyphs = glyphs;
        this.alignment = normalized.alignment.compose(new Alignment(alignment));
    }
}

/**
 * Infer the alignment between two strings with a "smart" heuristic.
 *
 * We use Unicode normalization and case mapping to minimize differences that are due to case, accents, ligatures, etc.
 */
export default function heuristicInfer(original: string, modified: string): BiString {
    const augOrig = new AugmentedString(original);
    const augMod = new AugmentedString(modified);

    let alignment = Alignment.infer(augOrig.glyphs, augMod.glyphs, AugmentedGlyph.costFn);
    alignment = augOrig.alignment.compose(alignment);
    alignment = alignment.compose(augMod.alignment.inverse());

    return new BiString(original, modified, alignment);
}
